### ISSUE TITLE
New rule proposal: Mismatched Column Counts in INSERT and SELECT Statements

### ISSUE THREAD

**Issue by wong-codaio** [sqlfluff/sqlfluff#5515]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### Description

Introduce a new rule in our SQL linter to enhance code accuracy and integrity. The rule should issue a warning when there's a mismatch in the number of columns specified in an INSERT statement compared to those in the corresponding SELECT statement. This feature aims to prevent potential run-time errors and data inconsistencies by ensuring column counts are aligned in these critical SQL operations. It will significantly improve the robustness of our database interactions.

### Use case

SQLFluff should catch issues like this, where the # of columns between INSERT and SELECT mismatch.

```
INSERT INTO table1
(col1)
SELECT
    col1,
    col2
FROM table2;
```

This should NOT apply if:
1. `INSERT INTO <table>` does not have a column list.
2. `SELECT` is simply doing a `*` (Which is a different issue.)

### Dialect

I think this should be applicable to all.

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)
